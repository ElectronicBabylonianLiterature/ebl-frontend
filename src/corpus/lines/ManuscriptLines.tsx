import ListForm from 'common/List'
import { createManuscriptLine } from 'corpus/text'
import React from 'react'
import { produce } from 'immer'
import { Col, Form } from 'react-bootstrap'
import _ from 'lodash'
import ArrayInput from 'common/ArrayInput'
import Editor from 'editor/Editor'

function ManuscriptLineForm({ value, manuscripts, onChange, disabled }) {
  const handleChange = (property) => (event) =>
    onChange(
      produce(value, (draft) => {
        draft[property] = event.target.value
      })
    )
  const handleIdChange = (event) =>
    onChange(
      produce(value, (draft) => {
        draft.manuscriptId = Number(event.target.value)
      })
    )
  return (
    <Form.Row>
      <Form.Group as={Col} md={2} controlId={_.uniqueId('ManuscriptLine-')}>
        <Form.Label>Siglum</Form.Label>
        <Form.Control
          as="select"
          value={value.manuscriptId}
          onChange={handleIdChange}
        >
          <option value="0" disabled>
            --
          </option>
          {manuscripts.map((manuscript) => (
            <option key={manuscript.id} value={manuscript.id}>
              {manuscript.siglum}
            </option>
          ))}
        </Form.Control>
      </Form.Group>
      <Col as={Col} md={1}>
        <ArrayInput
          value={value.labels}
          separator={' '}
          onChange={(labels) =>
            onChange(
              produce(value, (draft) => {
                draft.labels = labels
              })
            )
          }
        >
          Side
        </ArrayInput>
      </Col>
      <Form.Group as={Col} md={1} controlId={_.uniqueId('ManuscriptLine-')}>
        <Form.Label>Line nr.</Form.Label>
        <Form.Control value={value.number} onChange={handleChange('number')} />
      </Form.Group>
      <Col>
        <label>Text</label>
        <Editor
          name={_.uniqueId('Text-')}
          value={value.atf}
          onChange={(atf) =>
            onChange(
              produce(value, (draft) => {
                draft.atf = atf
              })
            )
          }
          disabled={disabled}
        />
      </Col>
    </Form.Row>
  )
}

export function ManuscriptLines({ lines, manuscripts, onChange, disabled }) {
  return (
    <ListForm
      noun="manuscript"
      defaultValue={createManuscriptLine({})}
      value={lines}
      onChange={onChange}
    >
      {(line, onChange) => (
        <ManuscriptLineForm
          onChange={onChange}
          value={line}
          manuscripts={manuscripts}
          disabled={disabled}
        />
      )}
    </ListForm>
  )
}
