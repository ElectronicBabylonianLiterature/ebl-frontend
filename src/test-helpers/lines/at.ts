import {
  DivisionAtLine,
  ObjectAtLine,
  SurfaceAtLine,
  HeadingAtLine,
  DiscourseAtLine,
  SealAtLine,
} from 'transliteration/domain/line'

export const surface: SurfaceAtLine = {
  prefix: '@',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'obverse?!',
      value: 'obverse?!',
      type: 'ValueToken',
    },
  ],
  // eslint-disable-next-line @typescript-eslint/camelcase
  surface_label: {
    status: ['UNCERTAIN', 'CORRECTION'],
    surface: 'OBVERSE',
    text: '',
    abbreviation: 'o',
  },
  displayValue: 'obverse?!',
  type: 'SurfaceAtLine',
}

export const object: ObjectAtLine = {
  prefix: '@',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'object Stone wig',
      value: 'object Stone wig',
      type: 'ValueToken',
    },
  ],
  label: {
    status: [],
    object: 'OBJECT',
    text: 'stone wig',
    abbreviation: 'stone wig',
  },
  displayValue: 'object Stone wig',
  type: 'ObjectAtLine',
}

export const heading: HeadingAtLine = {
  prefix: '@',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'h1',
      value: 'h1',
      type: 'ValueToken',
    },
  ],
  number: 1,
  displayValue: 'h1',
  type: 'HeadingAtLine',
}

export const division: DivisionAtLine = {
  prefix: '@',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'm=division paragraph 1',
      value: 'm=division paragraph 1',
      type: 'ValueToken',
    },
  ],
  text: 'paragraph',
  number: 1,
  displayValue: 'm=division paragraph 1',
  type: 'DivisionAtLine',
}

export const discourse: DiscourseAtLine = {
  prefix: '@',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'colophon',
      value: 'colophon',
      type: 'ValueToken',
    },
  ],
  // eslint-disable-next-line @typescript-eslint/camelcase
  discourse_label: 'COLOPHON',
  displayValue: 'colophon',
  type: 'DiscourseAtLine',
}

export const seal: SealAtLine = {
  prefix: '@',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'seal 1',
      value: 'seal 1',
      type: 'ValueToken',
    },
  ],
  number: 1,
  displayValue: 'seal 1',
  type: 'SealAtLine',
}
