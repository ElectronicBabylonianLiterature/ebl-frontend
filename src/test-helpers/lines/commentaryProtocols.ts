import { TextLine } from 'transliteration/domain/line'

export const commentaryProtocols = new TextLine({
  prefix: '9.',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'kur',
      value: 'kur',
      language: 'AKKADIAN',
      normalized: false,
      lemmatizable: true,
      uniqueLemma: [],
      parts: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          name: 'kur',
          nameParts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              type: 'ValueToken',
            },
          ],
          subIndex: 1,
          modifiers: [],
          flags: [],
          sign: null,
          type: 'Reading',
        },
      ],
      type: 'Word',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '!bs',
      value: '!bs',
      type: 'CommentaryProtocol',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'kur',
      value: 'kur',
      language: 'AKKADIAN',
      normalized: false,
      lemmatizable: true,
      uniqueLemma: [],
      parts: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          name: 'kur',
          nameParts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              type: 'ValueToken',
            },
          ],
          subIndex: 1,
          modifiers: [],
          flags: [],
          sign: null,
          type: 'Reading',
        },
      ],
      type: 'Word',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '!cm',
      value: '!cm',
      type: 'CommentaryProtocol',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'kur',
      value: 'kur',
      language: 'AKKADIAN',
      normalized: false,
      lemmatizable: true,
      uniqueLemma: [],
      parts: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          name: 'kur',
          nameParts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              type: 'ValueToken',
            },
          ],
          subIndex: 1,
          modifiers: [],
          flags: [],
          sign: null,
          type: 'Reading',
        },
      ],
      type: 'Word',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '!qt',
      value: '!qt',
      type: 'CommentaryProtocol',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'kur',
      value: 'kur',
      language: 'AKKADIAN',
      normalized: false,
      lemmatizable: true,
      uniqueLemma: [],
      parts: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          name: 'kur',
          nameParts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              type: 'ValueToken',
            },
          ],
          subIndex: 1,
          modifiers: [],
          flags: [],
          sign: null,
          type: 'Reading',
        },
      ],
      type: 'Word',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '!zz',
      value: '!zz',
      type: 'CommentaryProtocol',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'kur',
      value: 'kur',
      language: 'AKKADIAN',
      normalized: false,
      lemmatizable: true,
      uniqueLemma: [],
      parts: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          name: 'kur',
          nameParts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              type: 'ValueToken',
            },
          ],
          subIndex: 1,
          modifiers: [],
          flags: [],
          sign: null,
          type: 'Reading',
        },
      ],
      type: 'Word',
    },
  ],
  lineNumber: {
    number: 9,
    hasPrime: false,
    prefixModifier: null,
    suffixModifier: null,
    type: 'LineNumber',
  },
  type: 'TextLine',
})
