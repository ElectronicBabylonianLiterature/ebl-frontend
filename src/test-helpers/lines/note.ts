import { NoteLine } from 'transliteration/domain/line'

const note: NoteLine = {
  prefix: '#note: ',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'this is a note ',
      value: 'this is a note ',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@i{italic text}',
      value: '@i{italic text}',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: ' ',
      value: ' ',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@akk{kur}',
      value: '@akk{kur}',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: ' ',
      value: ' ',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@sux{kur}',
      value: '@sux{kur}',
      type: 'ValueToken',
    },
  ],
  parts: [
    {
      text: 'this is a note ',
      type: 'StringPart',
    },
    {
      text: 'italic text',
      type: 'EmphasisPart',
    },
    {
      text: ' ',
      type: 'StringPart',
    },
    {
      language: 'AKKADIAN',
      tokens: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          language: 'AKKADIAN',
          normalized: false,
          lemmatizable: true,
          uniqueLemma: [],
          parts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              name: 'kur',
              nameParts: [
                {
                  enclosureType: [],
                  erasure: 'NONE',
                  cleanValue: 'kur',
                  value: 'kur',
                  type: 'ValueToken',
                },
              ],
              subIndex: 1,
              modifiers: [],
              flags: [],
              sign: null,
              type: 'Reading',
            },
          ],
          type: 'Word',
        },
      ],
      type: 'LanguagePart',
    },
    {
      text: ' ',
      type: 'StringPart',
    },
    {
      language: 'SUMERIAN',
      tokens: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          language: 'SUMERIAN',
          normalized: false,
          lemmatizable: false,
          uniqueLemma: [],
          parts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              name: 'kur',
              nameParts: [
                {
                  enclosureType: [],
                  erasure: 'NONE',
                  cleanValue: 'kur',
                  value: 'kur',
                  type: 'ValueToken',
                },
              ],
              subIndex: 1,
              modifiers: [],
              flags: [],
              sign: null,
              type: 'Reading',
            },
          ],
          type: 'Word',
        },
      ],
      type: 'LanguagePart',
    },
  ],
  type: 'NoteLine',
}

export default note
