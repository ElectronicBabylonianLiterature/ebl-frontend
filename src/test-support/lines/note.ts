import { NoteLine } from 'transliteration/domain/note-line'
import Reference from 'bibliography/domain/Reference'
import BibliographyEntry from 'bibliography/domain/BibliographyEntry'

export const note: NoteLine = new NoteLine({
  prefix: '#note: ',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'this is a note ',
      value: 'this is a note ',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@i{italic text}',
      value: '@i{italic text}',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: ' ',
      value: ' ',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@akk{kur}',
      value: '@akk{kur}',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: ' ',
      value: ' ',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@sux{kur}',
      value: '@sux{kur}',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@bib{RN1@5}',
      value: '@bib{RN1@5}',
      type: 'ValueToken',
    },
  ],
  parts: [
    {
      text: 'this is a note ',
      type: 'StringPart',
    },
    {
      text: 'italic text',
      type: 'EmphasisPart',
    },
    {
      text: ' ',
      type: 'StringPart',
    },
    {
      language: 'AKKADIAN',
      tokens: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          language: 'AKKADIAN',
          normalized: false,
          lemmatizable: true,
          uniqueLemma: [],
          parts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              name: 'kur',
              nameParts: [
                {
                  enclosureType: [],
                  erasure: 'NONE',
                  cleanValue: 'kur',
                  value: 'kur',
                  type: 'ValueToken',
                },
              ],
              subIndex: 1,
              modifiers: [],
              flags: [],
              sign: null,
              type: 'Reading',
            },
          ],
          type: 'Word',
        },
      ],
      type: 'LanguagePart',
    },
    {
      text: ' ',
      type: 'StringPart',
    },
    {
      language: 'SUMERIAN',
      tokens: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          language: 'SUMERIAN',
          normalized: false,
          lemmatizable: false,
          uniqueLemma: [],
          parts: [
            {
              enclosureType: [],
              erasure: 'NONE',
              cleanValue: 'kur',
              value: 'kur',
              name: 'kur',
              nameParts: [
                {
                  enclosureType: [],
                  erasure: 'NONE',
                  cleanValue: 'kur',
                  value: 'kur',
                  type: 'ValueToken',
                },
              ],
              subIndex: 1,
              modifiers: [],
              flags: [],
              sign: null,
              type: 'Reading',
            },
          ],
          type: 'Word',
        },
      ],
      type: 'LanguagePart',
    },
    {
      reference: {
        id: 'RN1',
        type: 'DISCUSSION',
        pages: '5',
        notes: '',
        linesCited: [],
      },
      type: 'BibliographyPart',
    },
  ],
  type: 'NoteLine',
})

export const hydratedNote: NoteLine = new NoteLine({
  prefix: '#note: ',
  content: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@bib{RN1@5}',
      value: '@bib{RN1@5}',
      type: 'ValueToken',
    },
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: '@bib{RN2@3}',
      value: '@bib{RN3@3}',
      type: 'ValueToken',
    },
  ],
  parts: [
    {
      reference: {
        id: 'RN1',
        type: 'DISCUSSION',
        pages: '5',
        notes: '',
        linesCited: [],
      },
      type: 'BibliographyPart',
    },
    {
      reference: new Reference(
        'DISCUSSION',
        '3',
        '',
        [],
        new BibliographyEntry({
          id: 'RN2',
          title: 'Title',
          type: 'article-journal',
          issued: {
            'date-parts': [[2020, 5, 18]],
          },
          volume: '12',
          page: '1-9',
          issue: 7,
          'container-title': 'Container',
          author: [{ given: 'Alice', family: 'Bob' }],
          URL: 'http://example.com',
        })
      ),
      type: 'BibliographyPart',
    },
  ],
  type: 'NoteLine',
})

export default note
