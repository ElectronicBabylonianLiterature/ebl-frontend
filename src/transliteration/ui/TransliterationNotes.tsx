import React from 'react'
import classNames from 'classnames'
import { NoteLine } from 'transliteration/domain/note-line'
import { MarkupPart } from 'transliteration/domain/markup'
import { noteNumber } from 'transliteration/domain/text'
import {
  isLanguagePart,
  isBibliographyPart,
} from 'transliteration/domain/type-guards'
import { LinkToLine } from './note-links'
import {
  DisplayBibliographyPart,
  DisplayLaguagePart,
  DisplayTextPart,
} from './markup'

function DisplayNoteLine({
  line,
  container = 'div',
}: {
  line: NoteLine
  container?: string
}): JSX.Element {
  return React.createElement(
    container,
    { className: classNames([`Transliteration__${line.type}`]) },
    line.parts.map((part: MarkupPart, index: number) => {
      if (isLanguagePart(part)) {
        return <DisplayLaguagePart key={index} part={part} />
      } else if (isBibliographyPart(part)) {
        return <DisplayBibliographyPart key={index} part={part} />
      } else {
        return <DisplayTextPart key={index} part={part} />
      }
    })
  )
}

export default function TransliterationNotes({
  notes,
}: {
  notes: ReadonlyMap<number, readonly NoteLine[]>
}): JSX.Element {
  return (
    <ol className="Transliteration__notes">
      {Array.from(notes).flatMap(([lineNumber, lines]) =>
        lines.map((line, index) => {
          const number = noteNumber(notes, lineNumber, index)
          return (
            <li key={`${lineNumber}-${index}`} id={`note-${number}`}>
              <LinkToLine lineNumber={lineNumber} number={number} />
              <DisplayNoteLine container="span" line={line} />
            </li>
          )
        })
      )}
    </ol>
  )
}
