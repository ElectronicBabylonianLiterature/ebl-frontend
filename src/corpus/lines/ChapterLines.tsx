import React from 'react'
import { Button, Col, Form } from 'react-bootstrap'
import _ from 'lodash'
import ListForm from 'common/List'
import Editor from 'editor/Editor'
import { createDefaultLineFactory } from 'corpus/line-factory'
import { produce } from 'immer'
import { ManuscriptLines } from './ManuscriptLines'

function ChapterLineForm({ value, manuscripts, onChange, disabled }) {
  const handleChangeValue = (property) => (propertyValue) =>
    onChange(
      produce(value, (draft) => {
        draft[property] = propertyValue
      })
    )
  const handleChange = (property) => (event) =>
    onChange(
      produce(value, (draft) => {
        draft[property] = event.target.value
      })
    )
  return (
    <>
      <Form.Row>
        <Form.Group as={Col} md={1} controlId={_.uniqueId('Lines-')}>
          <Form.Label>Number</Form.Label>
          <Form.Control
            value={value.number}
            onChange={handleChange('number')}
          />
        </Form.Group>
        <Col>
          <label>Ideal reconstruction</label>
          <Editor
            name={_.uniqueId('IdealReconstruction-')}
            value={value.reconstruction}
            onChange={handleChangeValue('reconstruction')}
            disabled={disabled}
          />
        </Col>
      </Form.Row>
      <ManuscriptLines
        lines={value.manuscripts}
        manuscripts={manuscripts}
        onChange={handleChangeValue('manuscripts')}
        disabled={disabled}
      />
    </>
  )
}

export default function ChapterLines({ chapter, onChange, onSave, disabled }) {
  const handleChange = (lines) =>
    onChange(
      produce(chapter, (draft) => {
        draft.lines = lines
      })
    )
  return (
    <Form>
      <fieldset disabled={disabled}>
        <ListForm
          noun="line"
          defaultValue={createDefaultLineFactory(_.last(chapter.lines))}
          value={chapter.lines}
          onChange={handleChange}
        >
          {(line, onChange) => (
            <ChapterLineForm
              onChange={onChange}
              value={line}
              manuscripts={chapter.manuscripts}
              disabled={disabled}
            />
          )}
        </ListForm>
        <Button onClick={onSave}>Save lines</Button>
      </fieldset>
    </Form>
  )
}
