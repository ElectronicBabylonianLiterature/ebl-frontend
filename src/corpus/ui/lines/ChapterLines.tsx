import React from 'react'
import { Badge, Button, Col, Form } from 'react-bootstrap'
import _ from 'lodash'
import ListForm from 'common/List'
import Editor from 'editor/Editor'
import { createDefaultLineFactory } from 'corpus/application/line-factory'
import produce, { castDraft } from 'immer'
import { ManuscriptLines } from './ManuscriptLines'
import {
  Manuscript,
  Line,
  Chapter,
  LineVariant,
  createVariant,
} from 'corpus/domain/text'

interface VariantFormProps {
  value: LineVariant
  manuscripts: readonly Manuscript[]
  onChange: (line: LineVariant) => void
  disabled?: boolean
}

function LineVariantForm({
  value,
  manuscripts,
  onChange,
  disabled = false,
}: VariantFormProps) {
  const handleChange = (property: string) => (propertyValue): void =>
    onChange(
      produce(value, (draft) => {
        draft[property] = propertyValue
      })
    )

  return (
    <>
      <Form.Row>
        <Col>
          <label>Ideal reconstruction</label>
          <Editor
            name={_.uniqueId('IdealReconstruction-')}
            value={value.reconstruction}
            onChange={handleChange('reconstruction')}
            disabled={disabled}
          />
        </Col>
      </Form.Row>
      <ManuscriptLines
        lines={value.manuscripts}
        manuscripts={manuscripts}
        onChange={handleChange('manuscripts')}
        disabled={disabled}
      />
    </>
  )
}

interface FormProps {
  value: Line
  manuscripts: readonly Manuscript[]
  onChange: (line: Line) => void
  disabled?: boolean
}

function ChapterLineForm({
  value,
  manuscripts,
  onChange,
  disabled = false,
}: FormProps) {
  const handleChangeBoolean = (
    property: string,
    propertyValue: boolean
  ): void =>
    onChange(
      produce(value, (draft) => {
        draft[property] = propertyValue
      })
    )

  const handleNumberChange = (
    event: React.ChangeEvent<HTMLInputElement>
  ): void => {
    onChange(
      produce(value, (draft) => {
        draft.number = event.target.value
      })
    )
  }
  const handleChange = (variants: LineVariant[]): void =>
    onChange(
      produce(value, (draft) => {
        draft.variants = castDraft(variants)
      })
    )
  return (
    <>
      <Form.Row>
        <Form.Group as={Col} md={1} controlId={_.uniqueId('Lines-')}>
          <Form.Label>Number</Form.Label>
          <Form.Control value={value.number} onChange={handleNumberChange} />
        </Form.Group>
        <Col md={11}>
          <Form.Check
            inline
            type="checkbox"
            id={_.uniqueId('secondLineOfParallelism-')}
            label="second line of parallelism"
            checked={value.isSecondLineOfParallelism}
            onChange={(): void =>
              handleChangeBoolean(
                'isSecondLineOfParallelism',
                !value.isSecondLineOfParallelism
              )
            }
          />
          <Form.Check
            inline
            type="checkbox"
            id={_.uniqueId('beginningOfSection-')}
            label="beginning of a section"
            checked={value.isBeginningOfSection}
            onChange={(): void =>
              handleChangeBoolean(
                'isBeginningOfSection',
                !value.isBeginningOfSection
              )
            }
          />
        </Col>
      </Form.Row>
      <ListForm
        noun="variant"
        defaultValue={createVariant({})}
        value={value.variants}
        onChange={handleChange}
      >
        {(variant: LineVariant, onChange: (variant: LineVariant) => void) => (
          <LineVariantForm
            onChange={onChange}
            value={variant}
            manuscripts={manuscripts}
            disabled={disabled}
          />
        )}
      </ListForm>
    </>
  )
}

interface ChapterLinesLinesProps {
  chapter: Chapter
  onChange: (chapter: Chapter) => void
  onSave: () => unknown
  disabled?: boolean
}

export default function ChapterLines({
  chapter,
  onChange,
  onSave,
  disabled = false,
}: ChapterLinesLinesProps): JSX.Element {
  const handleChange = (lines: Line[]): void =>
    onChange(
      produce(chapter, (draft) => {
        draft.lines = castDraft(lines)
      })
    )
  return (
    <Form>
      <Badge variant="warning">Beta</Badge>
      <fieldset disabled={disabled}>
        <ListForm
          noun="line"
          defaultValue={createDefaultLineFactory(_.last(chapter.lines))}
          value={chapter.lines}
          onChange={handleChange}
        >
          {(line: Line, onChange: (line: Line) => void) => (
            <ChapterLineForm
              onChange={onChange}
              value={line}
              manuscripts={chapter.manuscripts}
              disabled={disabled}
            />
          )}
        </ListForm>
        <Button onClick={onSave}>Save lines</Button>
      </fieldset>
    </Form>
  )
}
