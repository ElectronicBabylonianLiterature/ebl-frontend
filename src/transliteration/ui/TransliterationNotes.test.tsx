import { render, RenderResult, within } from '@testing-library/react'
import React from 'react'
import { hydratedNote, note } from 'test-helpers/lines/note'
import TransliterationNotes from './TransliterationNotes'

let element: RenderResult
let lines: HTMLElement[]

beforeEach(() => {
  const notes = new Map([
    [0, [note]],
    [1, []],
    [2, [note, hydratedNote]],
  ])
  element = render(<TransliterationNotes notes={notes} />)
  lines = element.getAllByRole('listitem')
})

test('Snapshot', () => {
  expect(element.container).toMatchSnapshot()
})

test('Shows all notes', () => {
  expect(lines.length).toEqual(3)
})

describe.each([
  [1, 0],
  [2, 2],
  [3, 2],
])('Note %s from line %s', (noteNumber, lineNumber) => {
  let line: HTMLElement

  beforeEach(() => {
    line = lines[noteNumber - 1]
  })

  test('ID', () => {
    expect(line).toHaveAttribute('id', `note-${noteNumber}`)
  })

  test('link', () => {
    const link = within(line).getByRole('link')

    expect(link).toHaveAttribute('href', `#line-${lineNumber}`)
    expect(link).toHaveTextContent(String(noteNumber))
  })
})
