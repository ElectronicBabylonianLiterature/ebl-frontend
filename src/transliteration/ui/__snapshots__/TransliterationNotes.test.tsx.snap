// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <ol
    class="Transliteration__notes"
  >
    <li
      id="note-1"
    >
      <a
        class="Transliteration__NoteLink"
        href="#line-0"
      >
        1
      </a>
       
      <span>
        <span>
          this is a note 
        </span>
        <em>
          italic text
        </em>
        <span>
           
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--AKKADIAN"
        >
          <span
            class="word-info__wrapper"
          >
            <span
              class=""
            >
              <span
                class="Transliteration__Reading"
              >
                <span
                  class=""
                >
                  <span
                    class="Transliteration__ValueToken"
                  >
                    <span
                      class=""
                    >
                      kur
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
          </span>
        </span>
        <span>
           
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--SUMERIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--SUMERIAN"
        >
          <span
            class="word-info__wrapper"
          >
            <span
              class=""
            >
              <span
                class="Transliteration__Reading"
              >
                <span
                  class=""
                >
                  <span
                    class="Transliteration__ValueToken"
                  >
                    <span
                      class=""
                    >
                      kur
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
          </span>
        </span>
        <span
          class="badge badge-danger"
        >
          @bib{
          RN1
          @
          5
          }
        </span>
      </span>
    </li>
    <li
      id="note-2"
    >
      <a
        class="Transliteration__NoteLink"
        href="#line-2"
      >
        2
      </a>
       
      <span>
        <span>
          this is a note 
        </span>
        <em>
          italic text
        </em>
        <span>
           
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--AKKADIAN"
        >
          <span
            class="word-info__wrapper"
          >
            <span
              class=""
            >
              <span
                class="Transliteration__Reading"
              >
                <span
                  class=""
                >
                  <span
                    class="Transliteration__ValueToken"
                  >
                    <span
                      class=""
                    >
                      kur
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
          </span>
        </span>
        <span>
           
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--SUMERIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--SUMERIAN"
        >
          <span
            class="word-info__wrapper"
          >
            <span
              class=""
            >
              <span
                class="Transliteration__Reading"
              >
                <span
                  class=""
                >
                  <span
                    class="Transliteration__ValueToken"
                  >
                    <span
                      class=""
                    >
                      kur
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
          </span>
        </span>
        <span
          class="badge badge-danger"
        >
          @bib{
          RN1
          @
          5
          }
        </span>
      </span>
    </li>
    <li
      id="note-3"
    >
      <a
        class="Transliteration__NoteLink"
        href="#line-2"
      >
        3
      </a>
       
      <span>
        <span
          class="badge badge-danger"
        >
          @bib{
          RN1
          @
          5
          }
        </span>
        <span
          class="reference-popover__citation"
        >
          Bob, 2020: 3
          <span
            class="type-abbreviation"
          >
             (D)
          </span>
        </span>
      </span>
    </li>
  </ol>
</div>
`;
